### PR TITLE
MGDOBR-717 - [UI] added test for Instance List pagination

### PR DESCRIPTION
**Pull request by ChughShilpa**
https://issues.redhat.com/browse/MGDOBR-717

### CODE REVIEW

**Review by sjamboro — COMMENTED** [review]
I am not sure if we could use this approach. It is the first step to create something like Page Fragments which is not recommended by [Cypress - best practice](https://docs.cypress.io/guides/references/best-practices#Organizing-Tests-Logging-In-Controlling-State). The original websites shows plenty of examples how to deal with it. 

I agree that some isolated functions are good to share. IMHO ... delete and create Instance are such functions (two occurences). I think that we can use commands or actions or create separate file (something like Utils.ts). I am not sure if export function from a test file is a good idea. 

Such functions should be:
- parameterized (name, state, etc.). See my comments.
- assert the neccesary objects (does we always assert that the instance is in the appropriate state or wait that the state is changed or number of columns)

Probably large projects are using PageFragments because you can find blog articles like [Stop using Page Objects and Start using App Actions](https://www.cypress.io/blog/2019/01/03/stop-using-page-objects-and-start-using-app-actions/#page-objects-in-cypress).

File: cypress/integration/InstancesTest.ts
Comment (sjamboro) [inline cypress/integration/InstancesTest.ts:90]: Do not forget to delete these lines before you merge it

File: cypress/integration/InstancesTest.ts
Comment (sjamboro) [inline cypress/integration/InstancesTest.ts:97]: Try to avoid using generated ID. It is fine if you can use `[aria-label='.....']`.

File: cypress/integration/InstancesTest.ts
Comment (sjamboro) [inline cypress/integration/InstancesTest.ts:124]: Please unify to use (count => {})

File: cypress/integration/InstancesTest.ts
Comment (sjamboro) [inline cypress/integration/InstancesTest.ts:148]: This peace of code is hard to read.

File: cypress/integration/InstancesTest.ts
Comment (jomarko) [reply cypress/integration/InstancesTest.ts:3]: it would be better to have some `DeleteUtil` file, or maybe some general `Delete` cusotm cypress command, that would be able to delete anything? instance, source, action ?

**Comment by jomarko** [comment]
I will postpone my review once @sjamboro comments are addressed/responded.

**Comment by sjamboro** [comment]
> I will postpone my review once @sjamboro comments are addressed/responded.

@jomarko can you take a look at [Instance List pagination test](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/blob/deb8d3c05f7cadda1e17c6b82202c8782bf47cfc/cypress/integration/InstancesTest.ts#L89-L164). Does it what you expect?

**Review by jomarko — COMMENTED** [review]
@sjamboro yes, I think the added test brings reasonable level of checks to the pagination feature. 

However the code looks quite long. So I still wait for your comments to be addressed first to not make the review process 'messy'.

For now I puts just two suggestions how to make the assertions more robust. 

Thank you for the PR @ChughShilpa

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts]: Done

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts]: Done

File: cypress/integration/DeleteTest.ts
Comment (ChughShilpa) [reply cypress/integration/DeleteTest.ts]: I couldn't understand this statement. 
As per my understanding, we can delete Instances which are only in "Ready" State

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts]: @jomarko @sjamboro 
I created Util.ts file for common functions like CreateInstance, DeleteInstance.
For now, I accounted for Instances only but we can make this more generic like 'Create or Delete anything" in future

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts]: Bifurcated the code into different test cases and i hope it is easy to read now.

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts]: > I am not sure if we could use this approach. It is the first step to create something like Page Fragments which is not recommended by [Cypress - best practice](https://docs.cypress.io/guides/references/best-practices#Organizing-Tests-Logging-In-Controlling-State). The original websites shows plenty of examples how to deal with it.
> 
> I agree that some isolated functions are good to share. IMHO ... delete and create Instance are such functions (two occurences). I think that we can use commands or actions or create separate file (something like Utils.ts). I am not sure if export function from a test file is a good idea.
> 
> Such functions should be:
> 
> * parameterized (name, state, etc.). See my comments.
> * assert the neccesary objects (does we always assert that the instance is in the appropriate state or wait that the state is changed or number of columns)
> 
> Probably large projects are using PageFragments because you can find blog articles like [Stop using Page Objects and Start using App Actions](https://www.cypress.io/blog/2019/01/03/stop-using-page-objects-and-start-using-app-actions/#page-objects-in-cypress).



> d

Thanks @jomarko and @sjamboro for the guidance and suggestions :)

Worked on the suggested changes. I preferred creating "Utility file- Util.ts" for common functions ( create and delete instance)
and kept these function as generic one  - a parameterized function with asserting necessary objects only

and hence created different function (newInstanceStatus and deleteInstanceStatus) which basically acts as wait of their expected behaviour (like createInstance should be in ready state and deleteInstance should not exist)

Kindly let me know your opinion on this.

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts]: I tried creating a common method to count Instances to avoid code duplicate but as Cypress is asynchronous and it doesn't work simply with assigning a return value to a variable. 

This is my understanding so far.

File: cypress/integration/InstancesTest.ts
Comment (sjamboro) [inline cypress/integration/InstancesTest.ts:16]: @ChughShilpa is a problem to leave here this variable?

File: cypress/integration/Util.ts
Comment (sjamboro) [inline cypress/integration/Util.ts:18]: @ChughShilpa I am a little bit confused by name of this function. When I read code I was not sure about its purpose. I think about `waitInstanceIsReady(instanceName)` or `waitInstanceStatus(instanceName, status)`.
It would be nice to have timeout as optional parameter.

@jomarko what is your opinion?

File: cypress/integration/Util.ts
Comment (sjamboro) [inline cypress/integration/Util.ts:35]: @ChughShilpa REST api says that delete is possible only in "ready" or "failed" state. We should allow to define "failed" state here.

File: cypress/integration/Util.ts
Comment (sjamboro) [inline cypress/integration/Util.ts:49]: @ChughShilpa again I am a little bit confused by the name of this function. The name contains "Status" and the code checks that the instance does not exist.

**Review by sjamboro — COMMENTED** [review]
thanks @ChughShilpa ... I am ready to approved it when you give me answers on my minor comments and when you rebase on the top of the main branch and solve merge conflicts.

You cannot change code but give me a reason why you do it maybe I missed something.

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts:16]: @sjamboro  As now i created generic function for createInstance and thus passing the instance name as parameter while calling the same function. IMO, there's no need to explicitly assigned it to a constant.

File: cypress/integration/InstancesTest.ts
Comment (sjamboro) [inline cypress/integration/InstancesTest.ts:22]: @ChughShilpa, in case that the `newInstanceName` variable stay in code you cannot change this line.

File: cypress/integration/InstancesTest.ts
Comment (sjamboro) [inline cypress/integration/InstancesTest.ts:25]: @ChughShilpa, in case that the newInstanceName variable stay in code you cannot change this line.

File: cypress/integration/InstancesTest.ts
Comment (sjamboro) [reply cypress/integration/InstancesTest.ts:16]: @ChughShilpa I have created this variable because I needed its value several times and some occurences stays in code even if you created new function. 

Thank you for your answer.

File: cypress/integration/Util.ts
Comment (jomarko) [reply cypress/integration/Util.ts:18]: @sjamboro yes, your proposal improves the code readabiloity, +1 to follow your proposal

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts:16]: i reverted my changes and keep the code as it is

File: cypress/integration/Util.ts
Comment (ChughShilpa) [reply cypress/integration/Util.ts]: Renamed the function name as suggested

File: cypress/integration/Util.ts
Comment (ChughShilpa) [reply cypress/integration/Util.ts]: Added "failed" state

File: cypress/integration/Util.ts
Comment (ChughShilpa) [reply cypress/integration/Util.ts]: modified the function name

**Review by jomarko — CHANGES_REQUESTED** [review]
@ChughShilpa please fix your PR, you committed some merge conflicts strings and your code is not compiling now.

**Comment by ChughShilpa** [comment]
> @ChughShilpa please fix your PR, you committed some merge conflicts strings and your code is not compiling now.



> @ChughShilpa please fix your PR, you committed some merge conflicts strings and your code is not compiling now.

yes am looking into it

**Comment by ChughShilpa** [comment]
> @ChughShilpa please fix your PR, you committed some merge conflicts strings and your code is not compiling now.

Fixed the PR

File: cypress/integration/InstancesTest.ts
Comment (jomarko) [inline cypress/integration/InstancesTest.ts:90]: From my point of view this 'previous page' click is not needed, as we do not check the effect of the click action.

- We should either remove it
- Or we should add assertions like we have after clicking 'next page'

File: cypress/integration/InstancesTest.ts
Comment (jomarko) [inline cypress/integration/InstancesTest.ts:118]: Could we please ad assertion that `pageSize` is equal to 20?

**Review by jomarko — COMMENTED** [review]
Thank you for a PR @ChughShilpa

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts:90]: I firstly added assertions for current page (first page) as 
`cy.get("[aria-label='Go to first page']").should("be.disabled");
        cy.get("[aria-label='Go to previous page']").should("be.disabled");
        cy.get("[aria-label='Go to last page']").should("not.be.disabled");
        cy.get("[aria-label='Go to next page']").should("not.be.disabled");`
        
Then move to next page and asserted for next page as 
`// Go to next page
        cy.get("[aria-label='Go to next page']").first().click();
        cy.get("[aria-label='Go to first page']").should("not.be.disabled");
        cy.get("[aria-label='Go to previous page']").should("not.be.disabled");

        // as currently count of instances are covered in 2 pages
        cy.get("[aria-label='Go to last page']").should("be.disabled");
        cy.get("[aria-label='Go to next page']").should("be.disabled");`
        
and then move to previous page (which is the same as first page)

Comment (jomarko) [reply cypress/integration/InstancesTest.ts:90]: `and then move to previous page (which is the same as first page) for which we already did the validation`

I agree that previous page is same as first page, where you did some assertions already. However the operation (last line) in your test:
`cy.get("[aria-label='Go to previous page']").first().click();`

How are we sure that after that clic really the fist page is shown? Are we sure there is never ending loading screen? or some other issue?

Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts:90]: I agree with your on this point. Will add the assertions for the same

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts:118]: @jomarko 
UI shows the listing of instances as

1. 20 per page
image

2. 10 per page


Thus, pageSize should be equal to 'total no of instances' not pageSize

Comment (jomarko) [reply cypress/integration/InstancesTest.ts:118]: +1

File: cypress/integration/InstancesTest.ts
Comment (ChughShilpa) [reply cypress/integration/InstancesTest.ts]: Done

**Review by jomarko — APPROVED** [review]
Thank you for changes

**Pull request merged**
